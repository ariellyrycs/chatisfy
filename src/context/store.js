import React, { useReducer, createContext, useContext } from 'react';
import Mock from '../Mock';
import { generateUId, generateTime } from '../Utils/General.js';


const currentUser = {
	id: generateUId(),
	nickName: ''
};

const Context = createContext();

const useAppContext = () => useContext(Context);

const initialState = {
	...Mock,
	currentUser
};

const mainReduce = (state, action) => {
	switch (action.type) {
		case 'UPDATE_ROOM':
			return {
				...state,
				selectedRoom: action.roomId
			};
		case 'SEND_MESSAGE':
			let currentId = generateUId();
			let rooms = state.rooms;
			let messageByRooms = rooms.messages;
			let messages = state.messages;
			let currentRoomId = state.selectedRoom,
				sentBy = action.sentBy;
			return {
				...state,
				rooms: {
					...rooms,
					messages: {
						...rooms.messages,
						[currentRoomId]: [...(messageByRooms[currentRoomId] || []), currentId]
					}
				},
				messages: {
					...messages,
					data: {
						...messages.data,
						[currentId]: {
							id: currentId,
							from: sentBy,
							content: action.message,
							createdTime: generateTime(),
							seenBy: [],
							own: action.from === sentBy
						}
					}
				}
			};
		case 'UPDATE_NICKNAME':
			return {
				...state,
				currentUser: {
					...state.currentUser,
					nickName: action.nickName
				}
			};
		default:
			return state;
	}
};

function ContextProvider(props) {
	const [state, dispatch] = useReducer(mainReduce, initialState);
	return (
		<Context.Provider value={{ state, dispatch }}>
			{props.children}
		</Context.Provider>
	);
};

export { ContextProvider, Context, useAppContext };