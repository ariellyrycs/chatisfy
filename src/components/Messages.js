import React, { useRef, useLayoutEffect } from 'react';
import { useAppContext } from '../context/store';

import { Grid, Box, Fab, useScrollTrigger, Zoom } from '@material-ui/core';

import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown';
import { makeStyles } from '@material-ui/styles';
import clsx from 'clsx';

const useStyles = makeStyles(() => {
    return {
        goDownButton: {
            position: 'absolute',
            right: '45px',
            bottom: '0px'
        },
        chatView: {
            height: 'calc(100vh - 204px)',
            overflow: 'scroll'
        },
        messageWrap: {
            backgroundColor: '#DCDCDC',
            margin: '1px 2px',
            padding: '2px 5px',
            borderRadius: '5px',
            maxWidth: '70%',
            float: 'left'
        },
        own: {
            backgroundColor: 'rgb(220, 248, 198)',
            float: 'right'
        },
        time: {
            color: 'grey',
            fontSize: '10px',
            textAlign: 'right'
        }
    }
});


const ScrollBottom = ({ children, container, bottonReference, chatRoomNode }) => {
    const classes = useStyles();
    const threshold = 100;
    const topTreshhold = useScrollTrigger({
        target: container,
        threshold: threshold
    });
    //This needs to be done to revert the orientation of useScrollTrigger
    const bottomTreshhold = !topTreshhold && container && chatRoomNode.offsetHeight > container.offsetHeight + threshold;

    const handleClick = () => {
        bottonReference.scrollIntoView({ behavior: 'smooth', block: 'center' });
        return false;
    };

    return (
        <Zoom in={bottomTreshhold}>
            <div onClick={handleClick} role='presentation' className={classes.goDownButton}>
                {children}
            </div>
        </Zoom>
    );
}

function MessagesContainer({ currentMessages, chatRoomNode, classes, currentUser, users }) {
    return (
        <Grid container justify='center' alignContent='flex-start' ref={chatRoomNode}>
            {currentMessages.map((message) =>
                <Grid key={message.id} item xs={11}>
                    <div className={clsx(classes.messageWrap, message.own ? classes.own : '')}>
                        <Box component='span'>{message.own ? currentUser.nickName : users.data[message.from].nickName}</Box>
                        <Box>{message.content}</Box>
                        <div className={classes.time}><time>{message.createdTime}</time></div>
                    </div>
                </Grid>
            )}
        </Grid>
    )
}

export default function Messages({ currentMessagesIds }) {
    const chatRoomNode = useRef();
    const { state } = useAppContext();
    const currentMessages = currentMessagesIds.map(id => state.messages.data[id]);
    const classes = useStyles();
    const container = useRef();
    const bottonReference = useRef();

    useLayoutEffect(() => {
        //This has to be done this way because because delays 
        bottonReference.current.scrollIntoView({ block: 'center' });
    }, [currentMessagesIds]);

    return (
        <div ref={container} className={classes.chatView}>
            <MessagesContainer
                classes={classes}
                chatRoomNode={chatRoomNode}
                currentMessages={currentMessages}
                currentUser={state.currentUser}
                users={state.users}
            />
            <ScrollBottom
                container={container.current}
                bottonReference={bottonReference.current}
                chatRoomNode={chatRoomNode.current}>
                <Fab color='secondary' size='small' aria-label='scroll down'>
                    <KeyboardArrowDownIcon />
                </Fab>
            </ScrollBottom>
            <div ref={bottonReference} />
        </div>
    );
}