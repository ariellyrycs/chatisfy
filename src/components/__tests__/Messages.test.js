import React from 'react';
import Messages from '../Messages';
import * as Store from '../../context/store';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'


describe('<Messages />', () => {
    test('render the message component', () => {

        const contextValues = {
            state: {
                users: {
                    1: {
                        name: 'GEORGE',
                        thumbnail: ''
                    },
                    2: {
                        name: 'SANDRA',
                        thumbnail: ''
                    },
                },
                messages: {
                    data: {
                        1: {
                            id: 1,
                            from: 1,
                            content: 'Make sure you don\'t, as they say, go whole',
                            createdTime: '12:12',
                            seen: true
                        },
                        2: {
                            id: 2,
                            from: 2,
                            content: 'Are you and Karen gonna, as they say, pig out?',
                            createdTime: '12:14',
                            seen: false
                        }
                    }
                }
            }
        };
        const currentMessagesIds = [1, 2];
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Messages currentMessagesIds={currentMessagesIds}/>);
        expect(toJson(tree)).toMatchSnapshot();
    });
});