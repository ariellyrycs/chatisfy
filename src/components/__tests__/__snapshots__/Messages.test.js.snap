// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Messages /> render the message component 1`] = `
<div
  className="makeStyles-chatView-2"
>
  <MessagesContainer
    chatRoomNode={
      Object {
        "current": undefined,
      }
    }
    classes={
      Object {
        "chatView": "makeStyles-chatView-2",
        "goDownButton": "makeStyles-goDownButton-1",
        "messageWrap": "makeStyles-messageWrap-3",
        "own": "makeStyles-own-4",
        "time": "makeStyles-time-5",
      }
    }
    currentMessages={
      Array [
        Object {
          "content": "Make sure you don't, as they say, go whole",
          "createdTime": "12:12",
          "from": 1,
          "id": 1,
          "seen": true,
        },
        Object {
          "content": "Are you and Karen gonna, as they say, pig out?",
          "createdTime": "12:14",
          "from": 2,
          "id": 2,
          "seen": false,
        },
      ]
    }
    users={
      Object {
        "1": Object {
          "name": "GEORGE",
          "thumbnail": "",
        },
        "2": Object {
          "name": "SANDRA",
          "thumbnail": "",
        },
      }
    }
  />
  <ScrollBottom>
    <WithStyles(ForwardRef(Fab))
      aria-label="scroll down"
      color="secondary"
      size="small"
    >
      <KeyboardArrowDownIcon />
    </WithStyles(ForwardRef(Fab))>
  </ScrollBottom>
  <div />
</div>
`;
